notation/notation.go

```diff
@@ -0,0 +1,314 @@
+
+If each and every signature associated with the reference fail the verification,
+then Verify will return `ErrorVerificationFailed` error along with an array
+of `VerificationOutcome`.
+
+# Callers can pass the verification plugin config in VerifyOptions.PluginConfig
+
+For more details on signature verification, see https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#signature-verification
+*/
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (Descriptor, []*VerificationOutcome, error) {
+	var verificationOutcomes []*VerificationOutcome
+	artifactRef := opts.ArtifactReference
+	artifactDescriptor, err := repo.Resolve(ctx, artifactRef)
+	if err != nil {
+		return Descriptor{}, nil, ErrorSignatureRetrievalFailed{Msg: err.Error()}
+	}
+
+	trustpolicyDoc, err := verifier.TrustPolicyDocument()
+	if err != nil {
+		return Descriptor{}, nil, ErrorNoApplicableTrustPolicy{Msg: err.Error()}
+	}
+	trustPolicy, err := trustpolicy.GetApplicableTrustPolicy(trustpolicyDoc, artifactRef)
+	if err != nil {
+		return Descriptor{}, nil, ErrorNoApplicableTrustPolicy{Msg: err.Error()}
+	}
+	// ignore the error since we already validated the policy document
+	verificationLevel, _ := trustpolicy.GetVerificationLevel(trustPolicy.SignatureVerification)
+	if verificationLevel.Name == trustpolicy.LevelSkip.Name {
+		verificationOutcomes = append(verificationOutcomes, &VerificationOutcome{VerificationLevel: verificationLevel})
+		return Descriptor{}, verificationOutcomes, nil
+	}
+	opts.VerificationLevel = verificationLevel
+
+	// get signature manifests
+	var success bool
+	var verifiedSigBlobDesc ocispec.Descriptor
+	err = repo.ListSignatures(ctx, artifactDescriptor, func(signatureManifests []ocispec.Descriptor) error {
+		if len(signatureManifests) < 1 {
+			return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("no signatures are associated with %q, make sure the image was signed successfully", artifactRef)}
+		}
+		// if already verified successfully, no need to continue
+		if success {
+			return nil
+		}
+		// process signatures
+		for _, sigManifest := range signatureManifests {
+			// get signature envelope
+			sigBlob, sigBlobDesc, err := repo.FetchSignatureBlob(ctx, sigManifest)
+			if err != nil {
+				return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("unable to retrieve digital signature with digest %q associated with %q from the registry, error : %s", sigBlobDesc.Digest, artifactRef, err.Error())}
+			}
+			_, outcome, err := verifier.Verify(ctx, sigBlob, opts)
+			if err != nil {
+				if outcome != nil && outcome.Error != nil {
+					outcome.SignatureBlobDescriptor = &sigBlobDesc
+					verificationOutcomes = append(verificationOutcomes, outcome)
+				}
+				continue
+			}
+			outcome.SignatureBlobDescriptor = &sigBlobDesc
+			verificationOutcomes = append(verificationOutcomes, outcome)
```

Comment:
Should this be done irrespective if `Verify` returned error or not?

Reply:
At this point, the signature verification has succeeded. This append is saving the signature with details.
For signatures that failed the verification, their outcomes are saved in the above step.